flatfs.go

```diff
@@ -654,17 +678,11 @@ func (fs *Datastore) doDelete(key datastore.Key) error {
 
 func (fs *Datastore) Query(q query.Query) (query.Results, error) {
 	prefix := datastore.NewKey(q.Prefix).String()
-	if (prefix != "/") ||
-		len(q.Filters) > 0 ||
-		len(q.Orders) > 0 ||
-		q.Limit > 0 ||
-		q.Offset > 0 ||
-		!q.KeysOnly ||
-		q.ReturnExpirations ||
-		q.ReturnsSizes {
-		// TODO this is overly simplistic, but the only caller is
-		// `ipfs refs local` for now, and this gets us moving.
-		return nil, errors.New("flatfs only supports listing all keys in random order")
+	if prefix != "/" {
+		// This datastore can't include keys with multiple components.
+		// Therefore, it's always correct to return an empty result when
+		// the user requests a filter by prefix.
```

Comment:
Log a warning message? Very likely indicates bad usage of the datastore which might happen silently to the application above.